From dbaa09e5222dcff4d91f18ba0623e31361916b12 Mon Sep 17 00:00:00 2001
From: Vinayakumar B <vinayakumarb@apache.org>
Date: Wed, 7 Oct 2015 14:29:34 +0530
Subject: [PATCH 1313/2748] HDFS-9159. [OIV] : return value of the command is
 not correct if invalid value specified in '-p
 (processor)' option (Contributed by nijel)

(cherry picked from commit 61b3547c462a4e51d0a31c0a8c8ce210e2d90df8)
(cherry picked from commit 2b6fe90764b5995cd9a34a99330db22be7c0fd36)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/OfflineImageViewerPB.java

Change-Id: I327846676a43f807a72016052fc71e007e69833b
---
 .../offlineImageViewer/OfflineImageViewerPB.java   |    4 ++++
 .../offlineImageViewer/TestOfflineImageViewer.java |    9 +++++++++
 2 files changed, 13 insertions(+)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/OfflineImageViewerPB.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/OfflineImageViewerPB.java
index fe05979..247c457 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/OfflineImageViewerPB.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/OfflineImageViewerPB.java
@@ -190,6 +190,10 @@ public static int run(String[] args) throws Exception {
                 new PrintStream(new WriterOutputStream(out)), delimiter, tempPath)) {
           writer.visit(new RandomAccessFile(inputFile, "r"));
         }
+      } else {
+        System.err.println("Invalid processor specified : " + processor);
+        printUsage();
+        return -1;
       }
       return 0;
     } catch (EOFException e) {
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java
index 2cbc9f1..1f0b37e 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/TestOfflineImageViewer.java
@@ -354,6 +354,15 @@ public void testPBDelimitedWriter() throws IOException, InterruptedException {
         new FileSystemTestHelper().getTestRootDir() + "/delimited.db");
   }
 
+  @Test
+  public void testInvalidProcessorOption() throws Exception {
+    int status =
+        OfflineImageViewerPB.run(new String[] { "-i",
+            originalFsimage.getAbsolutePath(), "-o", "-", "-p", "invalid" });
+    assertTrue("Exit code returned for invalid processor option is incorrect",
+        status != 0);
+  }
+
   private void testPBDelimitedWriter(String db)
       throws IOException, InterruptedException {
     final String DELIMITER = "\t";
-- 
1.7.9.5

